File: HomeWork/src/main/java/week05/Solution875.java
package week05;

public class Solution875 {
    int[] piles;
    int h;

    public int minEatingSpeed(int[] piles, int h) {
        this.piles = piles;
        this.h = h;

        long left = 0;
        long right = 0;
        for (int pile : piles) {
            right += pile;
        }

        while (left < right) {
            long mid = (left + right) / 2;

            if (check(mid)) {
                right = mid;
            } else {
                left = mid + 1L;
            }
        }
        return (int) right;

    }

    private boolean check(long k) {
        int use_hour = 0;

        for (int pile : piles) {
            if (k == 0)
                return false;
            long need_hour = (long) Math.ceil((double) pile / (double) k);
            use_hour += need_hour;

            if (use_hour > h)
                return false;

        }
        return use_hour <= h;
    }
}

File: BinarySearch/src/main/python/solution875.py
from typing import List
import math


class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        def check(piles, h, k):
            use_hour = 0

            for pile in piles:
                # while pile > k:
                #     use_hour += 1
                #     pile -= k
                if k == 0:
                    return False
                need_hour = math.ceil(pile / k)

                use_hour += need_hour

                if use_hour > h:
                    return False

            return use_hour <= h

        left = 0
        right = sum(piles)

        while left < right:
            mid = (left + right) // 2

            if check(piles, h, mid):
                right = mid
            else:
                left = mid + 1

        return right


s = Solution()
# piles = [312884470]
# h = 312884469
piles = [312884470]
h = 968709470
res = s.minEatingSpeed(piles=piles, h=h)
print(res)
